File: agent/executor/ob/cluster.go
/*
 * Copyright (c) 2024 OceanBase.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ob

import (
	"github.com/oceanbase/obshell/agent/constant"
	"github.com/oceanbase/obshell/agent/errors"
	"github.com/oceanbase/obshell/agent/executor/tenant"
	"github.com/oceanbase/obshell/agent/meta"
	"github.com/oceanbase/obshell/agent/repository/db/oceanbase"
	"github.com/oceanbase/obshell/agent/repository/model/bo"
)

// only for mysql
func GetObclusterSummary() (*bo.ClusterInfo, *errors.OcsAgentError) {
	var info bo.ClusterInfo

	// Get all tenant infos
	tenantOverviews, err := tenantService.GetTenantsOverViewByMode(constant.MYSQL_MODE)
	if err != nil {
		return nil, errors.Occur(errors.ErrUnexpected, err.Error())
	}
	for _, tenantOverview := range tenantOverviews {
		tenantInfo, err := tenant.GetTenantInfo(tenantOverview.TenantName)
		if err != nil {
			return nil, err
		}
		info.Tenants = append(info.Tenants, *tenantInfo)
	}

	// Set basic cluster info
	if err := observerService.GetOBParatemerByName("cluster", &info.ClusterName); err != nil {
		return nil, errors.Occur(errors.ErrUnexpected, err.Error())
	}
	if err := observerService.GetOBParatemerByName("cluster_id", &info.ClusterId); err != nil {
		return nil, errors.Occur(errors.ErrUnexpected, err.Error())
	}
	isCommunityEdition, err := obclusterService.IsCommunityEdition()
	if err != nil {
		return nil, errors.Occur(errors.ErrUnexpected, err.Error())
	}
	info.IsCommunityEdition = isCommunityEdition
	info.Status = oceanbase.OBStateShortMap[oceanbase.GetState()]
	version, err := obclusterService.GetObVersion()
	if err != nil {
		return nil, errors.Occur(errors.ErrUnexpected, err.Error())
	}
	info.ObVersion = version

	rootServers, err := obclusterService.GetAllZoneRootServers()
	if err != nil {
		return nil, errors.Occur(errors.ErrUnexpected, err.Error())
	}

	zones, err := obclusterService.GetAllZone()
	if err != nil {
		return nil, errors.Occur(errors.ErrUnexpected, err.Error())
	}

	serverResourceMap, err := obclusterService.GetAllObserverResourceMap()
	if err != nil {
		return nil, errors.Occur(errors.ErrUnexpected, err.Error())
	}

	allAgents, _ := agentService.GetAllAgentsDOFromOB()
	archMap := make(map[string]string)
	for _, agent := range allAgents {
		archMap[meta.NewAgentInfo(agent.Ip, agent.RpcPort).String()] = agent.Architecture
	}

	for _, zone := range zones {
		var zoneInfo bo.Zone
		zoneInfo.Name = zone.Zone
		zoneInfo.IdcName = zone.Idc
		zoneInfo.RegionName = zone.Region
		zoneInfo.InnerStatus = zone.Status
		if zoneInfo.InnerStatus == "ACTIVE" {
			zoneInfo.Status = "RUNNING"
		} else if zoneInfo.InnerStatus == "INACTIVE" {
			zoneInfo.Status = "SERVICE_STOPPED"
		}
		if rootServer, ok := rootServers[zone.Zone]; ok {
			zoneInfo.RootServer = rootServer.ToBO()
		}
		observers, err := obclusterService.GetOBServersByZone(zone.Zone)
		if err != nil {
			return nil, errors.Occur(errors.ErrUnexpected, err.Error())
		}

		for _, server := range observers {
			observerBo := server.ToBo()
			observerBo.Architecture = archMap[meta.NewAgentInfo(server.SvrIp, server.SvrPort).String()]
			if baseResourceStat, ok := serverResourceMap[meta.ObserverSvrInfo{
				Ip:   server.SvrIp,
				Port: server.SvrPort,
			}]; !ok {
				continue
			} else {
				observerBo.Stats.BaseResourceStats = baseResourceStat.ToBO()
				observerBo.Stats.FillExtendDiskStats()
				observerBo.Stats.Zone = zone.Zone
				observerBo.Stats.Ip = server.SvrIp
				observerBo.Stats.Port = server.SvrPort
			}
			zoneInfo.Servers = append(zoneInfo.Servers, observerBo)
		}
		info.Zones = append(info.Zones, zoneInfo)
	}

	// Set cluster stats
	for _, zone := range info.Zones {
		for _, server := range zone.Servers {
			info.Stats.Add(&server.Stats.BaseResourceStats)
		}
	}

	// Set tenant stats
	for _, tenant := range info.Tenants {
		tenantSysStatsMap, err := obclusterService.GetTenantMutilSysStat(tenant.Id, []int{SYS_STAT_CPU_USAGE_STAT_ID, SYS_STAT_MEMORY_USAGE_STAT_ID, SYS_STAT_MAX_CPU_STAT_ID, SYS_STAT_MEMORY_SIZE_STAT_ID})
		if err != nil {
			return nil, errors.Occur(errors.ErrUnexpected, err.Error())
		}

		var tenantResourceStat bo.TenantResourceStat
		tenantResourceStat.TenantId = tenant.Id
		tenantResourceStat.TenantName = tenant.Name
		var cpuUsage, maxCpu, memoryUsage, memorySize float64
		if sysStat, ok := tenantSysStatsMap[SYS_STAT_CPU_USAGE_STAT_ID]; ok {
			cpuUsage = float64(sysStat.Value)
		}
		if sysStat, ok := tenantSysStatsMap[SYS_STAT_MAX_CPU_STAT_ID]; ok {
			maxCpu = float64(sysStat.Value)
		}
		if sysStat, ok := tenantSysStatsMap[SYS_STAT_MEMORY_USAGE_STAT_ID]; ok {
			memoryUsage = float64(sysStat.Value)
		}
		if sysStat, ok := tenantSysStatsMap[SYS_STAT_MEMORY_SIZE_STAT_ID]; ok {
			memorySize = float64(sysStat.Value)
		}
		tenantResourceStat.CpuUsedPercent = cpuUsage / maxCpu * 100
		tenantResourceStat.MemoryUsedPercent = memoryUsage / memorySize * 100
		tenantDataDiskUsageMap, err := tenantService.GetTenantDataDiskUsageMap()
		if err != nil {
			return nil, errors.Occur(errors.ErrUnexpected, err.Error())
		}
		tenantResourceStat.DataDiskUsage = tenantDataDiskUsageMap[tenant.Id]
		info.TenantStats = append(info.TenantStats, tenantResourceStat)
	}
	return &info, nil
}

File: ob/frontend/src/models/cluster.ts
/*
 * Copyright (c) 2024 OceanBase.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { obclusterInfo } from '@/service/obshell/obcluster';
import { useRequest } from 'ahooks';
import { flatten } from 'lodash';
import { useState } from 'react';

interface ClusterState {
  clusterListData: API.ClusterInfo[];
  tenantListData: API.TenantInfo[];
  zoneListData: API.Zone[];
  serverListData: (API.Observer & { zoneName?: string })[];
  clusterData: API.ClusterInfo;
  isSingleMachine: boolean;
  isStandalone: boolean;
  isCommunityEdition: boolean;
  isDistributedBusiness: boolean;
  isSharedStorage: boolean;
}

const clusterModel = () => {
  const [clusterState, setClusterState] = useState<ClusterState>({
    clusterListData: [],
    tenantListData: [],
    zoneListData: [],
    serverListData: [],
    clusterData: {} as API.ClusterInfo,
    isSingleMachine: false,
    isStandalone: false,
    isCommunityEdition: false,
    isDistributedBusiness: false,
    isSharedStorage: false, // 是否共享存储模式
  });

  const { run: getObclusterInfo, refresh, loading } = useRequest(obclusterInfo, {
    manual: true,
    onSuccess: res => {
      if (res.successful) {
        const clusterData = res?.data || ({} as API.ClusterInfo);
        // 是否单机版判断，如果clusterData里面zones只有一条数据，而且zones里面的servers只有一条数据，则认为单集群
        const isSingleMachine =
          clusterData.zones?.length === 1 && clusterData.zones[0].servers?.length === 1;

        // 计算是否为单机版
        const isStandalone = clusterData?.is_standalone || false;
        // 获取是否为社区版
        const isCommunityEdition = clusterData?.is_community_edition || false;
        // 获取是否是分布式商业版
        const isDistributedBusiness = !isStandalone && !isCommunityEdition;
        // 是否共享存储模式
        const isSharedStorage = clusterData?.is_shared_storage ?? false;

        const tenantListData = clusterData.tenants || [];
        const zoneListData = clusterData.zones || [];
        const serverListData: (API.Observer & { zoneName?: string })[] = flatten(
          zoneListData.map((item: API.Zone) =>
            (item.servers || []).map((server: API.Observer) => ({
              ...server,
              zoneName: item.name,
            }))
          )
        );

        setClusterState({
          clusterData,
          clusterListData: [clusterData],
          isSingleMachine,
          isStandalone,
          isCommunityEdition,
          isDistributedBusiness,
          isSharedStorage,
          tenantListData,
          zoneListData,
          serverListData,
        });
      }
    },
  });

  return {
    ...clusterState,
    getObclusterInfo,
    refreshClusterData: refresh,
    clusterDataLoading: loading,
  };
};

export default clusterModel;
